File: 1345-jump-game-iv/1345-jump-game-iv.java
class Solution {
    public int minJumps(int[] arr) {
        HashMap<Integer, List<Integer>> valueToIndex = new HashMap<Integer, List<Integer>>();
        
        //construct value2index hashmap
        for(int i = 0; i < arr.length; i++){
            if (! valueToIndex.containsKey(arr[i])){
                ArrayList<Integer> temp = new ArrayList<Integer>();
                temp.add(i);
                valueToIndex.put(arr[i], temp);
            }else{
                valueToIndex.get(arr[i]).add(i);
            }
        }
        boolean[] visited = new boolean[arr.length];
        Deque<Integer> deque = new LinkedList<Integer>();
        deque.offer(0);
        visited[0] = true;
        int step_count = 0;
        //BFS
        while(!deque.isEmpty()){
            int size = deque.size();
            for(int i = 0; i < size; i++){
                int index = deque.poll();
                //add i-1, i+1
                if (index == arr.length - 1) return step_count;
                if(index - 1 >= 0 && !visited[index-1]){
                    deque.offer(index-1);
                    visited[index-1] = true;
                }
                if(index + 1 < arr.length && !visited[index+1]){
                    deque.offer(index+1);
                    visited[index+1] = true;
                }
                List<Integer> neighbors = valueToIndex.get(arr[index]);
                for(int idx : neighbors){
                    if (!visited[idx]){
                    //    if (idx == arr.length - 1) return step_count;
                        deque.offer(idx);
                        visited[idx] = true;
                    }
                }
                valueToIndex.get(arr[index]).clear();
                
            }
            step_count++;
        }
        return step_count;
    }
}

File: 1345-jump-game-iv/1345-jump-game-iv.py
class Solution:
    def minJumps(self, arr: List[int]) -> int:
        #bfs
        valToIdx = defaultdict(set)
        for idx, num in enumerate(arr):
            valToIdx[num].add(idx)
        
        visited = set()
        n = len(arr)
        #[idx, step]
        queue = deque([[0, 0]])
        visited.add(0)
     #   visitedVal.add(arr[0])
        
        
        while (queue):
            poppedIdx, curStep = queue.popleft()

            if poppedIdx == n - 1:
                return curStep
            
            #nxt
            if poppedIdx - 1 >= 0 and poppedIdx - 1 not in visited:
                queue.append([poppedIdx - 1, curStep + 1])
                visited.add(poppedIdx - 1)
            if poppedIdx + 1 < n and poppedIdx + 1 not in visited:
                queue.append([poppedIdx + 1, curStep + 1])
                visited.add(poppedIdx + 1)
            for nxt in valToIdx[arr[poppedIdx]]:
                if nxt not in visited:
                    visited.add(nxt)
                    queue.append([nxt, curStep + 1])
            valToIdx[arr[poppedIdx]] = set()
        
        return -1
                